### PR TITLE
[WIP] Rule: no-constant-expressions

### PR DESCRIPTION
**Pull request by denis-sokolov**
Important disclaimer: I am probably in over my head, please be gentle with your criticism. This rule is work in progress, and code has not been tidied at all. This PR is a request for comment.

I've seen #987 and figured - I can do basic static analysis on the expressions.
Well, I did, but, of course, it's a potentially very deep problem. I have tried to be very conservative and avoid false positives at the cost of missing true positives. 
Please see the test cases to see what the current implementation handles and what it doesn't handle.

The question is - should I pursue this further? Are you even open for something like this?
The next step that can be strived for is handling multiple subexpressions at the same time to be able to handle cases such as `a === a`,  `a === b && b === c && c !== a`, `a === 5 && a + 3 === 8` and others.

Thanks.

### CODE REVIEW

File: tests/lib/rules/no-constant-expressions.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        { code: "5 === (a || true)", args: 1 },
+        { code: "a === 'true'", args: 1 },
+        { code: "'long string' + 'another string'", args: 1 },
+        { code: "3 + (-a)", args: 1 }
+    ],
+    invalid: [
+        { code: "5 + 3", args: 1, errors: [{ message: "Expression always 8", type: "BinaryExpression" }] },
+        { code: "true && false", args: 1, errors: [{ message: "Expression always false", type: "LogicalExpression" }] },
+        { code: "a * 0 + 5;", args: 1, errors: [{ message: "Expression always 5", type: "BinaryExpression"}] },
+        { code: "a === a;", args: 1, errors: [{ message: "Expression always true", type: "BinaryExpression"}] },
```
Comment (michaelficarra) [inline tests/lib/rules/no-constant-expressions.js]: This is not always true, both for the obvious case of NaN and the non-obvious case of a side-effecting getter.

File: tests/lib/rules/no-constant-expressions.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        { code: "a !== a;", args: 1, errors: [{ message: "Expression always false", type: "BinaryExpression"}] },
+        { code: "a / a;", args: 1, errors: [{ message: "Expression always 1", type: "BinaryExpression"}] },
```
Comment (michaelficarra) [inline tests/lib/rules/no-constant-expressions.js]: Again, not true because of NaN.

**Comment by nzakas** [comment]
Interesting approach, though looks pretty expensive and the code is really hard to follow due to lack of comments. Also, you are doing your own traversal of the AST whereas ESLint is already doing a traversal, so we're paying that tax twice. We typically don't allow rules to do their own traversals.

Can you try running `npm run perf` both with and without your commit and post the results?

**Comment by denis-sokolov** [comment]
Once again, sorry for the code being a stream of consciousness, that's not my usual approach.

I am doing the traversal down from expressions. `no-constant-condition` does the same, although, obviously, it's doing it much less.

Performance without the commit:
```
CPU Speed is 1800 with multiplier 7500000
Performance Run #1:  2793.157971ms
Performance Run #2:  2601.715877ms
Performance Run #3:  2607.4878909999998ms
Performance Run #4:  2577.686498ms
Performance Run #5:  2556.938616ms
Performance budget ok:  2601.715877ms (limit: 4166.666666666667ms)
```

Performance with the commit:
```
CPU Speed is 1800 with multiplier 7500000
Performance Run #1:  3134.480725ms
Performance Run #2:  3170.740427ms
Performance Run #3:  3163.533719ms
Performance Run #4:  3115.697719ms
Performance Run #5:  3174.662402ms
Performance budget ok:  3163.533719ms (limit: 4166.666666666667ms)
```

I do assume that the rule will be costly performance-wise, especially if we add a more complex inter-sub-expression analysis. But I assume that that's the general cost of this feature and it's impossible to make it more effective. Is noting in the documentation that the rule is considerably costly not a good approach?

File: tests/lib/rules/no-constant-expressions.js (hunk 1 above)
Comment (denis-sokolov) [inline tests/lib/rules/no-constant-expressions.js]: Thank you for noting.
I do know about `NaN`, but I considered that regardless of that, the code `a === a` must be an error if authored that way. If anyone expects a `NaN`, they'd hopefully use `isNaN` or `Number.isNaN`.
As for the side-effecting getter, that's interesting. But with respect to different coding styles, I will say that I don't think I'd accept such code in my code reviews.
Given that the purpose of the linting rules is to ensure a certain code quality at the cost of forbidding certain expressions, I'm happy to forbid using `a === a` for the side effects.

Comment (michaelficarra) [inline tests/lib/rules/no-constant-expressions.js]: I use `x !== x` for a NaN test, and you should, too. `isNaN` checks that its parameter is a `NaN` value *after* trying to ToNumber it, which means that `isNaN(isNaN)` is `true`. `Number.isNaN` is only available in ES6+ environments, so it isn't portable yet.

Comment (denis-sokolov) [inline tests/lib/rules/no-constant-expressions.js]: Yeah, I know that. That's why I use a polyfill or `typeof x === "number" && isNaN(x)`. In my opinion `x === x` is a non-straightforward way to detect for `NaN`, it's confusing if you've never seen it before, and it could theoretically break if the language was extended by new capabilities, such as overriding equality operator or adding new "nullish" values.
We could disable the detection of `x === x`, but the possible benefit it provides by catching typo errors together with forcing an extra clarity for code seems beneficial to me.
But I don't mind if it's changed, I don't feel strongly about it.
It's likely the rule won't be included anyway. :)

**Comment by michaelficarra** [comment]
If you're going to be doing your own traversal, you might as well use [esmangle's partial evaluator](https://github.com/estools/esmangle/blob/master/lib/evaluator.js).

**Comment by denis-sokolov** [comment]
How does it compare to [`static-eval`](https://github.com/substack/static-eval)?

**Comment by michaelficarra** [comment]
I'm not sure. I've never used that module, but I've also had some negative experiences with modules from that particular author. There's also [half-and-half](https://github.com/int3/half-and-half).

**Comment by nzakas** [comment]
As it is, the rule is too slow to include - it accounts for nearly 1/6 of the run time on its own, and I don't think that's rational for what the information this rule is producing.

**Comment by denis-sokolov** [comment]
That's fair. Perhaps I'll release it as a plugin.

**Comment by denis-sokolov** [comment]
For the future reference, the plugin is located here:
https://github.com/denis-sokolov/eslint-plugin-no-constant-expressions
